pilot/pkg/serviceregistry/serviceentry/controller.go

```diff
@@ -264,6 +264,11 @@ func (s *Controller) workloadEntryHandler(old, curr config.Config, event model.E
 		}
 		instance := s.convertWorkloadEntryToServiceInstances(wle, services, se, &key, s.Cluster())
 		instancesUpdated = append(instancesUpdated, instance...)
+		if event == model.EventDelete {
+			s.serviceInstances.deleteServiceEntryInstances(namespacedName, key)
```

Comment:
isn't L285-L291 doing this clean?

Reply:
It's updating based on the current SE, however if the SE is changed later, we will not be able to clean up instances associated with the old SE

Let's say we have a SE(host: a), then a WE is added, it will update service instances based on the SE(host: a). 
After that, the SE hostname is changed to b, but the instances added above(host: a) can't get freed

The added unit test is an example, it will fail without this change

Reply:
>After that, the SE hostname is changed to b, but the instances added above(host: a) can't get freed

Not fully understand, when this happen, we so such clean job in `serviceEntryHandler`

Reply:
> when this happen, we so such clean job in `serviceEntryHandler`

Yes we [clean old instances by `instancesBySE`](https://github.com/istio/istio/blob/84d9f86021c64c65bc5cdc0b236da85d2f525907/pilot/pkg/serviceregistry/serviceentry/controller.go#L356-L359), but in `workloadEntryHandler` we didn't update `instancesBySE`, so the stale instances can't be correctly deleted in `serviceEntryHandler`

Reply:
kind of understand

Reply:
@dddddai Can you remove serviceInstancesStore.instancesBySE, i looked a little bit more and looks redundant

Reply:
I'm afraid not.. why do you think it's redundant?
We store instances per SE, so that we can clean up old instances on SE update, if we don't have `instancesBySE` then we will have to look up old instances with the old SE. It's basically a space-time tradeoff I think

Reply:
NVM, this is my coarse idea.